torch/distributed/tensor/parallel/loss_parallel.py

```diff
@@ -0,0 +1,402 @@
+        )
+        x = x.to(computation_dtype)
+        if x.numel() == 0:
+            shifted = x
+        else:
+            x_max = torch.amax(x, dim, keepdim=True)
+            x_max = funcol.all_reduce(
+                x_max, reduceOp=c10d.ReduceOp.MAX.name, group=(mesh, mesh_dim)
+            )
+            shifted = x - x_max
+        shifted_sumexp = torch.sum(torch.exp(shifted), dim, keepdim=True)
+        shifted_sumexp = funcol.all_reduce(
+            shifted_sumexp, reduceOp=c10d.ReduceOp.SUM.name, group=(mesh, mesh_dim)
+        )
+        shifted_logsumexp = torch.log(shifted_sumexp)
+        result = shifted - shifted_logsumexp
+        if not half_to_float:
+            result = result.to(result_dtype)
+        return result
+
+    spec = x._spec
+    mesh_dim = _find_all_reduce_mesh_dim(spec.placements, dim)
+    res = _log_softmax(x._local_tensor, dim, half_to_float, spec.mesh, mesh_dim)
+
+    output_tensor_meta = _propagate_tensor_meta(op_call, args, kwargs)
+
+    return DTensor(
+        res,
+        spec.mesh,
+        spec.placements,
+        shape=output_tensor_meta.shape,
+        dtype=output_tensor_meta.dtype,
+        requires_grad=res.requires_grad,
+        stride=output_tensor_meta.stride,
+    )
+
+
+def _log_softmax_backward_handler(
+    op_call: torch._ops.OpOverload,
+    args: Tuple[object, ...],
+    kwargs: Dict[str, object],
+) -> object:
+    grad_output = cast(DTensor, args[0])
+    output = cast(DTensor, args[1])
+    dim = cast(int, args[2])
+    input_dtype = cast(torch.dtype, args[3])
+
+    # NOTE: As explained below at _nll_loss_and_log_softmax_backward, the
+    # _log_softmax_backward_handler does not actually do any computation.
+    def _log_softmax_backward_no_computation(
+        grad_output: Tensor, output: Tensor, dim: int, input_dtype: torch.dtype
+    ):
+        # The log_softmax backward computation has been performed during
+        # nll_loss backward in _nll_loss_and_log_softmax_backward.
+        # grad_input = grad_output - torch.exp(output) * torch.sum(
+        #     grad_output, dim=dim, keepdim=True
+        # )
+        grad_input = grad_output
+
+        if grad_output.dtype != input_dtype:
+            grad_input = grad_input.to(input_dtype)
+        return grad_input
+
+    spec = grad_output._spec
+    res = _log_softmax_backward_no_computation(
+        grad_output._local_tensor, output._local_tensor, dim, input_dtype
+    )
+
+    assert spec.tensor_meta is not None
+    return DTensor(
+        res,
+        spec.mesh,
+        spec.placements,
+        shape=spec.tensor_meta.shape,
+        dtype=input_dtype,
+        requires_grad=res.requires_grad,
+        stride=spec.tensor_meta.stride,
+    )
+
+
+# NOTE: The implementation follows torch._decomp.decomposition._nll_loss_forward,
+# with customized communication inserted to perform distributed computation.
+def _nll_loss_forward(
```

Comment:
thanks for implementing these decomp functions! 

I think we should add a function that's similar to `nn.functional.cross_entropy_loss` in this file, and just call it `cross_entropy_loss`, which the exact same args/kwargs (maybe have optional mesh/mesh_dim args, and only run allreduce when these are present)

Then we should write a test to compare the `nn.functional.cross_entropy_loss` and the function we have, to assert numerics.

This function we wrote can also be public, meaning that if user want to use ours directly, they can do that

Reply:
That's certainly doable. Basically local `cross_entropy` just needs to call `_log_softmax` and `_nll_loss_forward`, and for each of them, we can toggle all_reduce depending on whether `mesh` & `mesh_dim` are present.

However, am I right that this local `cross_entropy` won't have backward support, because the all_reduce's cannot be handled? So is the point of having this to provide user with more efficient forward-only computation when they want to use loss parallel with plain torch.Tensor input?